packages/python/openverse-attribution/src/openverse_attribution/license.py

```diff
@@ -1,128 +1,271 @@
-from enum import StrEnum
+import re
+from dataclasses import dataclass
 
-from openverse_attribution.attribution import get_attribution_text
+from openverse_attribution.data.all_licenses import all_licenses
+from openverse_attribution.license_name import LicenseName
 
 
-class License(StrEnum):
-    """
-    Represent all licenses that are handled by Openverse. This uses a very loose
-    interpretation of the term "license" as it includes licenses (both active
-    and deprecated), dedications and marks.
-    """
+KNOWN_ALIASES = {
+    "zero": "cc0",
+    "mark": "pdm",
+}
 
-    # CC licenses
-    BY = "by"
-    BY_SA = "by-sa"
-    BY_NC = "by-nc"
-    BY_ND = "by-nd"
-    BY_NC_SA = "by-nc-sa"
-    BY_NC_ND = "by-nc-nd"
 
-    # Deprecated CC licenses
-    SA = "sa"
-    SAMPLING = "sampling+"
-    NC_SAMPLING = "nc-sampling+"
+@dataclass
+class License:
+    name: LicenseName
+    ver: str | None
+    jur: str | None
 
-    # Public domain dedication
-    CC0 = "cc0"
+    def __init__(
+        self,
+        slug: str,
+        ver: str | None = None,
+        jur: str | None = None,
+    ):
```

Comment:
This constructor is a little confusing. I can pass a jurisdiction, even if the license doesn't support it :thinking: for example:

```py
License(slug="pdm", ver="4.0", jur="jp")
```

This is something meaningless about this constructor, but it would silently pass.

I think there is a way to validate this without needing to use this imperative code, either, and instead rely on the `all_licenses` to construct ccREL formatted references for all known licenses and marks. Flattening `all_licenses.json` into a set of strings like `{"by-nc/3.0/pl", "by-nc/2.0/jp", "by/4.0", ...}` and then either taking those strings directly as arguments (which would make it easier to support additional license namespaces in the future, without a significant API change) or as parts, in the current implementation, but construct the string and check for it in the flatted all_licenses.

We can also alias version with `*` to the latest jurisdiction, so `by-nc/*/jp` would map to `by-nc/2.0/jp` so that if the version isn't provided, fall back to `*` and it still gets mapped to the right place.

This would make most of this code declarative, a bit easier to follow/reason about, and would make supporting additional license namespaces trivial. To do that, we'd add a new `namespace` parameter in the vein of RDF namespaces, so libreart would have `https://artlibre.org/licence`, CC would have `https://creativecommons.org/licenses`, and so forth. That would just key to the mapping used, rather than needing to change the logic of this function.

It's worth also taking a look at some prior art, and the API of the library for parsing spdx signifiers: https://pypi.org/project/spdx-lookup/. In particular, the `by_id` and `by_name` methods, which are essentially similar to what I'm suggesting as far as using a single string argument. Validation is much easier this way, especially if we consider multiple license namespaces beyond CC.

Reply:
> This is something meaningless about this constructor, but it would silently pass.

It does not. It accepts partial info (for when version and jurisdiction are unknown) and completes it if possible. If the information is invalid it will not pass.

<img width="1317" alt="Screenshot 2024-05-02 at 8 36 38 AM" src="https://github.com/WordPress/openverse/assets/16580576/d1fa1bfc-faa2-4952-a766-1362166a7ca8">

> We can also alias version with * to the latest jurisdiction, so by-nc/*/jp would map to by-nc/2.0/jp so that if the version isn't provided, fall back to * and it still gets mapped to the right place.

The constructor already works like this. You can skip the fields you do not know. 

If only one match fits, it will populate it...

<img width="537" alt="Screenshot 2024-05-02 at 8 41 21 AM" src="https://github.com/WordPress/openverse/assets/16580576/bb1a7891-5bb5-4710-b621-d8adac3c4a74">

... else it will use the latest match as a `fallback`.

<img width="571" alt="Screenshot 2024-05-02 at 8 39 25 AM" src="https://github.com/WordPress/openverse/assets/16580576/a2a092e4-7fce-4ab9-b24e-d3e77c70a079">

Reply:
Apologies for misunderstanding the validation! Your reasoning sounds good to me :+1: